### ISSUE TITLE
Logitech MX Keys mini for Mac "fn" key not working

### ISSUE THREAD

**Issue by JackHansonC** [pqrs-org/Karabiner-Elements#3070]
I tried many times, but it simply does not work. When karabiner is turned on, and even I did nothing to the Fn key, it still does not function. For example, I set up in the system keyboard setting and set fn key as "show emoji", when I press fn, nothing happens. But if I map say right option to fn, then right option works as fn properly. Anyone shares similar problems?

**Comment by elcoda** [comment]
The same here, in Karabiner-eventViewer no event is traced when I press "fn" from my Logitech MX Mini.
If I use the "fn" key of my macbook pro (m1 silicon) event viewer shows the right keyboard_fn key

**Comment by basavyr** [comment]
I've got the same issue with my MX Keys Mini (non-mac version). I'm using Karabiner-Elements `v14.4.0` and Monterey `12.4`. Changing the **fn** seems impossible

**Comment by NguyenThien201** [comment]
Same here, any solutions?

**Comment by MuhammedZakir** [comment]
Without Karabiner-Elements, does `fn` key work? Quit KE and see if EventViewer detects `fn`. If it does not, then your keyboard does not send `fn` key event.

FWIW, generally, only Apple-keyboards and keyboards that identifies themselves as Apple-keyboards send `fn` key.

**Comment by p2p3p4** [comment]
MX Keys M Mac on MBP OSX 10.15.7 "fn" key reassignment doesn't seem to work (after 30 years finger memory refuses to forget that bottom leftmost key is not Control) Hoped to be able to move fn key action to the CapsLock key, make the original fn key a duplicate Control key.

Assigning "fn" to CapsLock key does seem to work.

Oddly, after the above assignments are made (and no others) top row (still on MX Keys for Mac kbd) the audio key assignments seem to have shifted one position left. Ie. Mute not does volume down, volume down does volume up.

**Comment by AliHoussein** [comment]
Same here. It looks like Logitech software, Logi Options has an option (actually that you can't opt-out) to use the Fn key for : "The Fn Key of your keyboard allow you to expand the functionality of your Logitech mouse."  
Do you have Logi Options installed?

**Comment by HEY-SAI** [comment]
Same problem with MX Keys mini.

**Comment by j08lue** [comment]
Seems like this issue is the same as described in the docs here: https://karabiner-elements.pqrs.org/docs/help/troubleshooting/touch-bar-function-keys/

So what we have here:
* KE running -> `fn` key on Logitech MX does not work at all
* KE off -> `fn` key works as expected (i.e. fn + media key gives corresponding function key).

I tried the solution from the docs but it did not change anything.

Btw, is this not a duplicate of https://github.com/pqrs-org/Karabiner-Elements/issues/2012?

🤷

**Comment by HEY-SAI** [comment]
I believe the first, cause the emoji keyboard not popping out when pressing fn button. but it works for mac keyboard

**Comment by nbasov** [comment]
Same issue with Logitech MX Keys on macOS Ventura with Karabiner 14.10.0 on Apple M1 Max. Played with @j08lue `s solution but no luck. MX's FN+F... keys  works perfect if this keyboard disabled in `Karabiner -> Devices`. If it is enabled, media keys stops working. Also there are no events in Event Viewer for them.

**Comment by nonameplum** [comment]
Please vote and observe https://support.logi.com/hc/pl/community/posts/9588846190871-How-can-I-swap-fn-and-ctrl-keys-on-MX-Mini-keys- . I think there is no other way around than Logi will provide a way in their software/firmware to switch the `fn` key.

**Comment by owenzhao** [comment]
I think this is one of the common issues of membrane keyboards. Unless mechanical keyboards, which directly output F1~F12 and Fn+F1~F12. Membrane keyboards output shortcuts and F1~F12 only, there are no Fn+F1~F12. That is why you can't get Fn key from viewer. 

So comparing to mechanical keyboards, with membrane keyboards, you can only get F1~F12 working with Karabiner by default. If you wants more, you have to manually convert shortcuts in custom one by one. But even if you do so, the solution is not perfect, as some of the shortcuts are not detect by viewer, which means Karabiner doesn't recognize the key or the key won't be send in Mac mode.

**Comment by carstenbauer** [comment]
For me, the solution was the following:
* Open Karabiner-Elements Settings.
* Go to "Function Keys" on the left.
* Activate the "Use all F1, F2, etc. keys as standard function keys" toggle (it was deactivated before).

Afterwards, `fn` + function key gives me, say, `f1`. Without `fn` the function keys still work as desired (e.g. brightness control).

(Note that I see the same weird mappings (`f1 -> ...`, `f2 -> ...`, etc.) as mentioned by @j08lue but I just left them as is.)

**Comment by feiyeung** [comment]
It'd be interesting to see if the `Fn` key itself has a key code sent to the host first. I don't have the keyboard to check, but it'd be interesting for someone with that and say a linux environment to see if `xev` acknowledges a keycode or not when `Fn` itself is pressed. Myself isn't familiar with Mac's driver or Karabiner to tell if the event viewer on Mac will do enough/equivalent to `xev` of linux. 

Reason asking is, on most Windows/PC laptop or keyboards designed for Windows/PC, the `Fn` key is filtered by the embedded controller (EC) without ever reporting to the OS. `Fn` + something combination is also interpreted by the EC, which then sends the corresponding event to OS for that combination.

If MX Keys (mini?) non-Mac version is having this same implementation, where the `Fn` key is not reported to the OS, then this is a firmware limitation on the keyboard, and i'd imagine there's nothing we can do to actually key filter the `Fn`, but to use workarounds for mapping the media keys.

**Comment by ukaytac** [comment]
This is still not working, any news?

**Comment by kflu** [comment]
> where the `Fn` key is not reported to the OS, then this is a firmware limitation on the keyboard, and i'd imagine there's nothing we can do to actually key filter the `Fn`, but to use workarounds for mapping the media keys

Yes you are correct. I tried this on Linux with `xev`. The `fn` key on this keyboard does not report as any key press. So it is really a key interpreted by the keyboard's firmware.

**Comment by andrew-trotter** [comment]
As others have stated, the fn key doesn't send a signal. My solution (on a macbook) was to remap right opt (control for windows) -> fn. It's a key I hardly use and it's right next to the fn key. That way I still get the functionality of the fn modifier.

**Comment by sethwilsonUS** [comment]
Same issue here. :(

**Comment by georgiyrr** [comment]
Same with MX keys mini for Mac

**Comment by maxkolenko** [comment]
Fn key on these Logitech keyboards is not supposed to send any keycode to operating system. It's made purposedly this way. Therefore you can't remap it. Thank you, Logitech.

**Comment by dmuiX** [comment]
such a weird thing. a third party program can do sth what the logitech sofware cannot do...
karabiner can activate the fn keys logi option+ cannot...

**Comment by nejmeijer** [comment]
Closing this as a duplicate of #2012, feel free to reopen if it is not.

**Comment by ananduush** [comment]
I used fn keys mostly for changing the input source (i use 3 input source). Assigning the emoji quick action was having some issues which displayed the wrong language in the small blue icon. After some digging, changing the input source via the keyboard shortcuts to change it via ctrl + space was the best solution for me.

**Comment by nejmeijer** [comment]
@ananduush Please refrain from commenting on closed issues marked as duplicate.

**Comment by belkocik** [comment]
So what is the key code of 'fn' key in mx keys s?

![Image](https://github.com/user-attachments/assets/3e2ddc83-1dae-4afe-82b4-82d627100910)

Key Codes app doesn't recognize 'fn' key. :/

**Comment by nejmeijer** [comment]
Thanks for the information @richardrebeco, re-opening again.

**Comment by Vladormir** [comment]
Still nothing ?
